data_and_training/unsloth-training/train_unsloth.py

```diff
@@ -0,0 +1,275 @@
+            args=TrainingArguments(
+            evaluation_strategy="steps",
+            do_eval=True,
+            save_strategy="steps",
+            eval_steps=eval_steps,
+            save_steps=save_steps,
+            per_device_train_batch_size=batch_size,
+            per_device_eval_batch_size=batch_size,
+            auto_find_batch_size=auto_batch_size,
+            # gradient_accumulation_steps = 4,
+            warmup_steps=int(len(train_ds) / 8),  # 5,
+            # max_steps = 60,
+            learning_rate=learning_rate,
+            fp16=not is_bfloat16_supported(),
+            bf16=is_bfloat16_supported(),
+            logging_steps=5,
+            optim="adamw_8bit",
+            weight_decay=weight_decay,
+            lr_scheduler_type = lr_scheduler,
+            # seed = random_state,
+            output_dir=output_name,
+            num_train_epochs=epochs,
+            load_best_model_at_end=True,
+        ),
+    )
+
+    gpu_stats = torch.cuda.get_device_properties(0)
+    start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
+    max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
+    print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
+    print(f"{start_gpu_memory} GB of memory reserved.")
+
+    trainer_stats = trainer.train()
+
+    used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
+    used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
+    used_percentage = round(used_memory / max_memory * 100, 3)
+    lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
+    print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
+    print(f"{round(trainer_stats.metrics['train_runtime'] / 60, 2)} minutes used for training.")
+    print(f"Peak reserved memory = {used_memory} GB.")
+    print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
+    print(f"Peak reserved memory % of max memory = {used_percentage} %.")
+    print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")
+
+    load_dotenv()
+    hf_token = os.getenv('HF_TOKEN')
+
+    # model.save_pretrained_merged(f'{output_name}_merged_4bit', tokenizer, save_method = "merged_4bit_forced", token=hf_token)
+    # model.push_to_hub_merged(f'{output_name}_merged_4bit', tokenizer, save_method="merged_4bit_forced", token=hf_token)
+
+    model.save_pretrained(f'models/{output_name}_local')  # Local saving
+    tokenizer.save_pretrained(f'models/{output_name}_local')
+
+    model.push_to_hub(f'{output_name}_lora', token=hf_token)  # Online saving
+    tokenizer.push_to_hub(f'{output_name}_lora', token=hf_token)  # Online saving
+
+    # model.save_pretrained_merged(f'{output_name}_lora', tokenizer, save_method = "lora", token=hf_token)
+    # model.push_to_hub_merged(f'{output_name}_lora', tokenizer, save_method = "lora", token=hf_token)
```

Comment:
I wonder if this was the issue of why we can not load the model :/ maybe we use these methods too.